game-core/src/test/java/games/strategy/engine/random/MartiDiceRollerTest.java

```diff
@@ -0,0 +1,82 @@
+package games.strategy.engine.random;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.collection.IsArray.array;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URI;
+import java.util.Arrays;
+import org.junit.jupiter.api.Test;
+
+public class MartiDiceRollerTest {
+  private static final String successMessage =
+      "<!DOCTYPE html>\n"
+          + "<html>\n"
+          + "\t<head>\n"
+          + "\t\t<title>M.A.R.T.I. Server -- \"more accurate rolls than irony\"</title>\n"
+          + "\t</head>\n"
+          + "\t<body>\n"
+          + "\t\t<p>Dice results were sent via email.</p>"
+          + "<br><a href='https://dice.marti.triplea-game.org/MARTI_verify.php?iv=oTVwKksuDf4f0LW4j"
+          + "6U6yAklQWsxpdac%2FmX1NMP7uXY%3D&enc=vhDa64l%2FORX6X8di6LVdl5JQGC81uQyKE4Jt0gHoVsOV5Uxz"
+          + "msbVReOsBQonZmQi7JZAwGhL4HoFXzkJMjilNX1gRMs5%2FhXyjyoUryYGPzgOWIVG7IxkuOMRiot7kPqJQwNc"
+          + "f08ev8TNj9WxZTSWPNy2j1Y17piXRn%2FffXIGPDDcrmw2rSeB3lCq6ZjlKSmMCIqEXk%2FSsJjQO7jKq7CkWT"
+          + "ZPY2QxTfFuOmFEj1Ji7pp1GVuVcc5g90q5'>Click here to verify the roll.</a>"
+          + "<br>your dice are: 2,5,1<p><p>\t</body>\n"
+          + "</html>";
+
+  private static final String failureMessage =
+      "<!DOCTYPE html>\n"
+          + "<html>\n"
+          + "\t<head>\n"
+          + "\t\t<title>M.A.R.T.I. Server -- \"more accurate rolls than irony\"</title>\n"
+          + "\t</head>\n"
+          + "\t<body>\n"
+          + "\t\tfatal error: Emails [user@mail.invalid] are not registered."
+          + " Please register them at https://dice.marti.triplea-game.org/register.php .!\n"
+          + "\n";
+
+  private final MartiDiceRoller martiDiceRoller =
+      MartiDiceRoller.builder()
+          .diceRollerUri(URI.create("http://uri.invalid"))
+          .ccAddress("")
+          .toAddress("")
+          .gameId("")
+          .build();
+
+  @Test
+  @SuppressWarnings("unchecked")
+  void successfulMessageExtraction() throws Exception {
+    final Integer[] dice =
+        Arrays.stream(martiDiceRoller.getDice(successMessage, 0)).boxed().toArray(Integer[]::new);
+
+    assertThat(dice, is(array(equalTo(1), equalTo(4), equalTo(0))));
+  }
+
+  @Test
+  void unsuccessfulMessageExtraction() {
+    final Exception exception =
+        assertThrows(
+            InvocationTargetException.class, () -> martiDiceRoller.getDice(failureMessage, 0));
+    assertThat(
+        exception.getMessage(),
+        is(
+            " Emails [user@mail.invalid] are not registered."
+                + " Please register them at https://dice.marti.triplea-game.org/register.php ."));
+    assertThat(exception.getCause(), is(nullValue()));
```

Comment:
Why is this assertion useful? Is it needed for correctness that the exception thrown have no cause?

Side-note, it's really questionable already we are using exception for control-flow handling. The exception being thrown makes it hard to tell if we are seeing a case where the dice-roller is down, no network, or the totally expected case where a user is not yet registered.

Reply:
Not really, it only sort-of ensures that this isn't some generic `InvocationTargetException` that is actually thrown for some legitimate reason.
I mean I'm convinced the designers of `InvocationTargetException` made the cause a mandatory argument for a reason in the constructor, in 99% of all cases there should be a cause

Reply:
Hmm, isn't 99% of all cases a user not being registered? It seems like almost everyone hits that case and before this exception message *was* the thing to tell you to register. It was a common problem that people would not grok the error message and were just left confused how to make the system work.

Overall, I think the root problem is we are dealing with control-flow-by-exception-handling anti-pattern.

If we are verifying those criteria, we can add messages to the assertion to document the reasoning and perhaps check the instance of the exception as well.

Reply:
Do you think it would help using an unchecked custom exception and throwing that and just catching a generic RuntimeException somewhere else?
In this case neither the exception message nor any exception causes would be relevant

Reply:
My thinking, at a low level, there should be something that does the send test and returns a payload that has an error message or a success flag. Any exception would then be for cases where we can't connect or cannot send. If there is a difference in how we would recover from such a scenario, then different exceptions would be appropriate.

Catching unchecked exceptions, or throwing a custom exception and then doing a catch-all to catch any runtime exception does not seem super easy to grok. Exceptions are effectively glorified GOTO statements after-all. 

Getting back to this test case, I think the things to do, to make the test clear, would be:
1. add an assertion comment to explain why we are verifying the condition. You stated so much in response to my questions earlier: "it only sort-of ensures that this isn't some generic InvocationTargetException that is actually thrown for some legitimate reason."

2. Perhaps also verify the instance type of the exception is an `InvocationTargetException`

Reply:
@DanVanAtta I re-did some of the Exceptions.
Basically explicit server error messages like "wrong/missing email" etc. now have their own exception type and the code has to explicitly catch that.
Semantic Errors like the server returning an invalid string that doesn't match the pattern is now treated as `IllegalStateException`, never explicitly caught and is supposed to be a likely programmer error.
All kind of network related exceptions are now treaded as such. As user-fixable errors and provides a link to the forum in case the error doesn't go away.

Let me know what you think, I tested all 3 scenarios (by changing the source code to invalid servers/URLs and entering non existing emails) and I believe the error pop-up seems appropriate in each case and provides the right amount of information to debug the issue.